### PR TITLE
Warn when using shouldBe between different types

### PR DESCRIPTION
**Pull request by Kantis**

### CODE REVIEW

File: src/test/resources/inspections/nullableUnequalTypes.kt [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import io.kotest.core.script.test
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.matchers.shouldBe
+
+class NullableUnequalTypesTest : FunSpec({
+
+   test("foo") {
+      val x: Int? = null
+      x shouldBe 5
+      5 shouldBe x
```
Comment (Kantis) [inline src/test/resources/inspections/nullableUnequalTypes.kt:12]: comparing Int? and Int should not cause warnings

**Comment by sksamuel** [comment]
Is this ready for review ?

**Comment by Kantis** [comment]
Conceptually, sure. I think I'd like to build and run the plugin with the changes for a while before claiming it's good to go for real tho. Just haven't gotten around to it yet.. 🙈

**Comment by sksamuel** [comment]
Ok makes sense.

**Comment by Kantis** [comment]
Managed to install it and played around with it for a bit.. Seems like it can be really useful to me 😄 

I also tried to make sure it didn't break performance. I opened some of the heavier files in the kotest and it didn't feel unusually slow to me. Tested with `TagsTest` and `StringMatchersTest`, since they to p usage of `shouldBe`, and `DoubleMatchersTest` since it's just big in general..

**Comment by sksamuel** [comment]
Is it ready for review and test ?

**Comment by Kantis** [comment]
I'm sorry, I discovered a couple of problems last night. I ran the inspection on Kotest and Hoplite sources and learnt that `shouldBe` supports matching `T` with `Matcher<T>`. Also, any pair of types that have `equals` implementations would probably give false alarms. Would be easy to not give any warnings when either type is `Matcher`, but not sure how to deal with the 2nd case..

**Comment by sksamuel** [comment]
What about keeping it simple.
If RHS is Matcher, don't run inspection
If RHS is a primitive or String, then the types must match.
Otherwise, LHS <:< RHS or RHS <:< LHS where <:< means subtype or type

**Comment by Kantis** [comment]
Thanks for the input, makes sense.. 
Fixed the Matcher handling (by simply checking if RHS is a type named `*Matcher*`)
Inspection currently passes for primitives which I think kotest handles comparison of gracefully (Int and Long, Double and Float, along with their respective primitive arrays). Should I ditch that?

**Comment by sksamuel** [comment]
I don't think ints and longs compare in kotest.

**Comment by Kantis** [comment]
@sksamuel this test passes for me, so it feels like they do? Am I missing something?
```
      test("dummy") {
         val i: Int = 6
         i shouldBe 6L
      }
```

**Comment by sksamuel** [comment]
I thought it failed when I tried.
What about 6L shouldBe 2 ?

**Comment by Kantis** [comment]
```

      test("dummy 1") {
         6L shouldBe 2
      }

-->
expected:<2L> but was:<6L>
Expected :2L
Actual   :6L



      test("dummy 2") {
         val y: Int = 2
         6L shouldBe y
      }
-->
expected:<2> but was:<6L>
Expected :2
Actual   :6L
```

**Comment by sksamuel** [comment]
There you go.

**Comment by Kantis** [comment]
I dont get it.. It's failing when we have different values, but passing when the same number? Feels like it should be considered fine then to compare ints and longs?

**Comment by sksamuel** [comment]
Oh wait, it's always going to fail on different values LOL.

Yeah you're probably right,

What about 6L shouldBe 6 ?

**Comment by Kantis** [comment]
```
      test("dummy 1") {
         6L shouldBe 6
      } --> pass

      test("dummy 2") {
         val y: Int = 6
         6L shouldBe y
      } --> pass
```

**Comment by sksamuel** [comment]
Ok then I guess it's fine to say int/long compare.

**Comment by Kantis** [comment]
Awesome :) do you have any hunch as to why the tests fail on just IC-212? 🤔

**Comment by sksamuel** [comment]
Nope, perhaps it's just not setup right, that was a recent addition to the
build matrix.

**Comment by Kantis** [comment]
> Nope, perhaps it's just not setup right, that was a recent addition to the build matrix.

Doh, so I just had to update those plugin.xml files 😂 sorry. Should be ok now.

File: src/main/kotlin/io/kotest/plugin/intellij/inspections/ShouldBeWithUnequalTypesInspection.kt [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            }
+         }
+      }
+   }
+
+   /**
+    * Fetches the type of [this] while ignoring nullability.
+    * E.g. if this is a `[Boolean]?` we will get [Boolean].
+    */
+   private fun KtExpression.tryResolveType(): KotlinType? =
```
Comment (sksamuel) [inline src/main/kotlin/io/kotest/plugin/intellij/inspections/ShouldBeWithUnequalTypesInspection.kt:46]: Can we move Psi helpers out to the psi helper package.

File: src/main/kotlin/io/kotest/plugin/intellij/inspections/ShouldBeWithUnequalTypesInspection.kt [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      val testedOperations = listOf("shouldBe", "shouldNotBe")
+
+      private val comparableTypes = setOf(
+         object : TypeSet("Int", "Long") {},
+         object : TypeSet("IntArray", "LongArray") {},
+         object : TypeSet("Float", "Double") {},
+         object : TypeSet("FloatArray", "DoubleArray") {},
+      )
+
+      abstract class TypeSet(vararg types: String) {
```
Comment (sksamuel) [inline src/main/kotlin/io/kotest/plugin/intellij/inspections/ShouldBeWithUnequalTypesInspection.kt:85]: why is this abstract ?

File: src/test/resources/inspections/nullableUnequalTypes.kt (hunk 1 above)
Comment (sksamuel) [reply src/test/resources/inspections/nullableUnequalTypes.kt:12]: Yeah I think that's fine.

File: src/test/resources/inspections/unequalTypes.kt [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package inspections
+
+import io.kotest.core.script.test
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.matchers.shouldBe
+
+class UnequalTypesTest : FunSpec({
```
Comment (sksamuel) [inline src/test/resources/inspections/unequalTypes.kt:7]: I would mark these as private unless you want them to run at CI level

File: src/test/resources/inspections/comparingWithMatcher.kt [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.matchers.Matcher
+import io.kotest.matchers.MatcherResult
+import io.kotest.matchers.shouldBe
+
+class MatcherTest : FunSpec({
+
+   test("foo") {
+      val x: Int? = null
+      x shouldBe AlwaysPassMatcher
```
Comment (sksamuel) [inline src/test/resources/inspections/comparingWithMatcher.kt:13]: Can we add a test for ComparableMatcher too

Comment (Kantis) [reply src/test/resources/inspections/comparingWithMatcher.kt:13]: Fixing 👍🏼

**Comment by Kantis** [comment]
@sksamuel thx for re-opening. 

I haven't gotten much feedback on this one so I'm still unsure if it has acceptable performance impact. I sometimes felt IntelliJ got a bit sluggish in large files, but not sure if it was due to these changes or just IntelliJ struggling. :)

**Pull request closed without merging**